File: DataPrep.R
################################################################################
# 1. LOAD REQUIRED PACKAGES
################################################################################

library(tidyverse)
library(readxl)
library(forcats)
library(dplyr)


################################################################################
# 2. IMPORT DATA
################################################################################

Data_to_clean <- read_excel("Data_Cleaning.xlsx")
Data_to_clean


################################################################################
# 3. SUMMARIZE VARIABLES WITH COUNT
################################################################################

Data_to_clean %>% count(Gender)
Data_to_clean %>% count(Age)
Data_to_clean %>% count(Status)


################################################################################
# 4. DATA CLEANING WORKFLOW
################################################################################

# Reload data
Data_to_clean <- read_excel("Data_Cleaning.xlsx")

# Remove rows with missing Status
DataCleaned <- Data_to_clean %>%
  filter(!is.na(Status))

# Remove duplicate rows
DataCleaned <- DataCleaned %>%
  distinct()

# Rename / correct gender levels
DataCleaned$Gender <- fct_recode(DataCleaned$Gender,
                                 "female" = "Female",
                                 "female" = "femail")

# Replace placeholder values (999) with NA in Age
DataCleaned$Age <- na_if(DataCleaned$Age, 999)

# View intermediate cleaned dataset
DataCleaned


################################################################################
# 5. HANDLE MISSING VALUES BY IMPUTATION
################################################################################

# Impute with mean
DataCleaned <- DataCleaned %>%
  mutate(Age = replace_na(Age, round(mean(Age, na.rm = TRUE))))

# Or impute with median
DataCleaned <- DataCleaned %>%
  mutate(Age = replace_na(Age, median(Age, na.rm = TRUE)))

# View final cleaned dataset
print(DataCleaned)


################################################################################
# 6. VISUALIZATION: COMPARATIVE BOX PLOT
################################################################################

p <- ggplot(DataCleaned, aes(Gender, Age))
p + geom_boxplot() + ggtitle("Comparative Boxplot: Gender vs Age")


################################################################################
# 7. ADDING A NEW VARIABLE (HEIGHT)
################################################################################

DataCleaned$Height_cm <- c(160, 175, 155, 165, 178, 190, 191, 180, 159, 167, 167, 159)
print(DataCleaned)


################################################################################
# 8. SCATTER PLOT: AGE VS HEIGHT
################################################################################

ggplot(DataCleaned, aes(x = Age, y = Height_cm)) +
  geom_point() +
  ggtitle("Scatter Plot: Age vs Height")


################################################################################
# 9. LOG TRANSFORMATION OF AGE
################################################################################

DataCleaned$Log_Age <- log(DataCleaned$Age)
print(DataCleaned)

# Scatter plot: Log Age vs Height
ggplot(DataCleaned, aes(x = Log_Age, y = Height_cm)) +
  geom_point() +
  ggtitle("Scatter Plot: Log(Age) vs Height")


################################################################################
# 10. SIMPLE LINEAR REGRESSION: HEIGHT ~ LOG(AGE)
################################################################################

lm.r <- lm(formula = Height_cm ~ Log_Age, data = DataCleaned)

# Display summary
summary(lm.r)


################################################################################
# 11. ADDING A SALARY VARIABLE
################################################################################

DataCleaned$Salary <- c(35, 22, 55, 25, 23, 20, 40, 45, 25, 19, 41)
print(DataCleaned)


################################################################################
# 12. CREATE A BOOTSTRAP SAMPLE
################################################################################

set.seed(123)  # for reproducibility

bootstrap_samples <- lapply(1:1000, function(i) {
  DataCleaned[sample(nrow(DataCleaned), nrow(DataCleaned), replace = TRUE), ]
})


################################################################################
# 13. BOOTSTRAPPING & LINEAR REGRESSION
################################################################################

boot_samples <- 1000  # Number of bootstrap samples
boot_results <- matrix(0, nrow = boot_samples, ncol = 2)

# Empty list to store bootstrap datasets
boot_data <- vector("list", boot_samples)

# Empty list to store models if you want them
boot_models <- vector("list", boot_samples)

for (i in 1:boot_samples) {
  
  # Bootstrap sample (same number of rows as original)
  samp <- sample(nrow(DataCleaned), nrow(DataCleaned), replace = TRUE)
  
  boot_data[[i]] <- DataCleaned[samp, ]
  
  # Fit MLR model (Salary ~ Age)
  mlr_model <- lm(Salary ~ Age, data = boot_data[[i]])
  
  # Store the model
  boot_models[[i]] <- mlr_model
  
  # Store coefficients (intercept + slope)
  boot_results[i, ] <- coef(mlr_model)
}

# =========================================
# 🔍 View the first bootstrap regression model
# =========================================
first_model <- boot_models[[1]]
summary(first_model)



################################################################################
# 14. MODEL EVALUATION
################################################################################

# Original predictor (Age, assuming column 3)
OrigDataSel <- DataCleaned[, 3]  

# Prepare matrices
n_boot <- nrow(boot_results)
n_obs <- nrow(DataCleaned)
Pred <- matrix(0, nrow = n_boot, ncol = n_obs)

# Compute predictions for each bootstrap model
for (i in 1:n_boot) {
  Pred[i, ] <- boot_results[i, 1] + boot_results[i, 2] * OrigDataSel
}

# Actual outcomes (Salary, assuming column 7)
Act <- matrix(rep(DataCleaned[, 7], each = n_boot), nrow = n_boot)

# Residuals
residuals <- Act - Pred

# RMSE for each bootstrap
RMSE <- sqrt(rowMeans(residuals^2))

# Visualize RMSE distribution
hist(RMSE, main = "Histogram of RMSE (Bootstrap)", col = "skyblue", border = "black")

# Normality test for RMSE
shapiro.test(RMSE)

################################################################################
# END OF SCRIPT
################################################################################

File: DataPrep.py
################################################################################
# 1. LOAD PACKAGES
################################################################################

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.utils import resample
import statsmodels.api as sm
from scipy.stats import shapiro

################################################################################
# 2. READ THE DATASET
################################################################################

# 2.1 Load Excel data
Data_to_clean = pd.read_excel("Data_Cleaning.xlsx")
print(Data_to_clean)

################################################################################
# 3. SUMMARIZE VARIABLES WITH COUNT
################################################################################

# 3.1 Count Gender
print(Data_to_clean['Gender'].value_counts())
# 3.2 Count Age
print(Data_to_clean['Age'].value_counts())
# 3.3 Count Status
print(Data_to_clean['Status'].value_counts())

################################################################################
# 4. DATA CLEANING WORKFLOW
################################################################################

# 4.1 Reload data
Data_to_clean = pd.read_excel("Data_Cleaning.xlsx")

# 4.2 Remove rows with missing Status
DataCleaned = Data_to_clean[~Data_to_clean['Status'].isna()]

# 4.3 Remove duplicate rows
DataCleaned = DataCleaned.drop_duplicates()

# 4.4 Correct spelling errors in Gender
DataCleaned['Gender'] = DataCleaned['Gender'].replace({
    'Female': 'female',
    'femail': 'female'
})

# 4.5 Replace placeholder (999) with NaN
DataCleaned['Age'] = DataCleaned['Age'].replace(999, np.nan)

################################################################################
# 5. HANDLE MISSING VALUES
################################################################################

# 5.1 Impute with mean
DataCleaned['Age'] = DataCleaned['Age'].fillna(round(DataCleaned['Age'].mean()))

# 5.2 Or impute with median (alternative)
# DataCleaned['Age'] = DataCleaned['Age'].fillna(DataCleaned['Age'].median())

print(DataCleaned)

################################################################################
# 6. BOX PLOT: GENDER VS AGE
################################################################################

sns.boxplot(x='Gender', y='Age', data=DataCleaned)
plt.title("Comparative Boxplot Gender vs Age")
plt.show()

################################################################################
# 7. ADD HEIGHT VARIABLE
################################################################################

DataCleaned['Height_cm'] = [160, 175, 155, 165, 178, 190, 191, 180, 159, 167, 167, 159]
print(DataCleaned)

################################################################################
# 8. SCATTER PLOT: AGE VS HEIGHT
################################################################################

sns.scatterplot(x='Age', y='Height_cm', data=DataCleaned)
plt.title("Scatter plot Age vs Height")
plt.show()

################################################################################
# 9. LOG TRANSFORMATION OF AGE
################################################################################

DataCleaned['Log_Age'] = np.log(DataCleaned['Age'])
print(DataCleaned)

# 9.1 Scatter plot: Log Age vs Height
sns.scatterplot(x='Log_Age', y='Height_cm', data=DataCleaned)
plt.title("Scatter plot Log Age vs Height")
plt.show()

################################################################################
# 10. SIMPLE LINEAR REGRESSION: HEIGHT ~ LOG_AGE
################################################################################

X = sm.add_constant(DataCleaned['Log_Age'])
y = DataCleaned['Height_cm']
lm_r = sm.OLS(y, X).fit()
print(lm_r.summary())

################################################################################
# 11. ADD SALARY VARIABLE
################################################################################

DataCleaned['Salary'] = [35, 22, 55, 25, 23, 20, 40, 45, 25, 19, 41, 22]
print(DataCleaned)

################################################################################
# 12. CREATE BOOTSTRAP SAMPLES
################################################################################

bootstrap_samples = [
    DataCleaned.sample(n=len(DataCleaned), replace=True)
    for _ in range(1000)
]

################################################################################
# 13. BOOTSTRAPPING & LINEAR REGRESSION
################################################################################

boot_samples = 1000
boot_results = np.zeros((boot_samples, 2))
boot_data = []
boot_models = []

n = len(DataCleaned)

for i in range(boot_samples):
    samp_idx = np.random.choice(n, n, replace=True)
    boot_df = DataCleaned.iloc[samp_idx].copy()
    boot_data.append(boot_df)

    X = sm.add_constant(boot_df["Age"])
    y = boot_df["Salary"]
    
    model = sm.OLS(y, X).fit()
    boot_models.append(model)

    boot_results[i, :] = model.params.values

# 13.1 View first bootstrap regression model
first_model = boot_models[0]
print(first_model.summary())

################################################################################
# 14. MODEL EVALUATION: RMSE
################################################################################

coef_list = []
rmse_list = []
r2_list = []

for i in range(boot_samples):
    boot_data_i = DataCleaned.sample(n=n, replace=True)
    
    X = boot_data_i[['Age']]  # example with Age only
    X = sm.add_constant(X)
    y = boot_data_i['Salary']
    
    model = sm.OLS(y, X).fit()
    
    coef_list.append(model.params.values)
    
    y_pred = model.predict(X)
    rmse = np.sqrt(np.mean((y - y_pred)**2))
    rmse_list.append(rmse)
    
    r2_list.append(model.rsquared)

boot_results_df = pd.DataFrame(coef_list, columns=['Intercept', 'Age'])
boot_results_df['RMSE'] = rmse_list
boot_results_df['R2'] = r2_list

plt.hist(boot_results_df['RMSE'], bins=30, color='skyblue', edgecolor='black')
plt.axvline(x=np.mean(boot_results_df['RMSE']), color='red', linewidth=2)
plt.title('Bootstrap RMSE Distribution')
plt.xlabel('RMSE')
plt.ylabel('Frequency')
plt.show()

plt.hist(boot_results_df['R2'], bins=30, color='lightgreen', edgecolor='black')
plt.title('Bootstrap R² Distribution')
plt.xlabel('R²')
plt.ylabel('Frequency')
plt.show()

################################################################################
# End of Script
################################################################################

